File: javascript/accounts-merge.js
// @leetup=custom
// @leetup=info id=721 lang=javascript slug=accounts-merge

/*
 * Given a list of `accounts` where each element `accounts[i]` is a list of
 * strings, where the first element `accounts[i][0]` is a name, and the rest of the
 * elements are emails representing emails of the account.
 *
 * Now, we would like to merge these accounts. Two accounts definitely belong to
 * the same person if there is some common email to both accounts. Note that even
 * if two accounts have the same name, they may belong to different people as
 * people could have the same name. A person can have any number of accounts
 * initially, but all of their accounts definitely have the same name.
 *
 * After merging the accounts, return the accounts in the following format: the
 * first element of each account is the name, and the rest of the elements are
 * emails in sorted order. The accounts themselves can be returned in any
 * order.
 *
 * Example 1:
 *
 * Input: accounts = [
 *   ["John","johnsmith@mail.com","john_newyork@mail.com"],
 *   ["John","johnsmith@mail.com","john00@mail.com"],
 *   ["Mary","mary@mail.com"],["John","johnnybravo@mail.com"]
 * ]
 * Output: [
 *   ["John","john00@mail.com","john_newyork@mail.com","johnsmith@mail.com"],
 *   ["Mary","mary@mail.com"],["John","johnnybravo@mail.com"]
 * ]
 * Explanation:
 * The first and second John's are the same person as they have the common email "j
 * ohnsmith@mail.com".
 * The third John and Mary are different people as none of their email addresses ar
 * e used by other accounts.
 * We could return these lists in any order, for example the answer [['Mary', 'mary
 * @mail.com'], ['John', 'johnnybravo@mail.com'],
 * ['John', 'john00@mail.com', 'john_newyork@mail.com', 'johnsmith@mail.com']] woul
 * d still be accepted.
 *
 * Example 2:
 *
 * Input: accounts = [
 *   ["Gabe","Gabe0@m.co","Gabe3@m.co","Gabe1@m.co"],
 *   ["Kevin","Kevin3@m.co","Kevin5@m.co","Kevin0@m.co"],
 *   ["Ethan","Ethan5@m.co","Ethan4@m.co", "Ethan0@m.co"],
 *   ["Hanzo","Hanzo3@m.co","Hanzo1@m.co","Hanzo0@m.co"],
 *   ["Fern","Fern5@m.co","Fern1@m.co","Fern0@m.co"]
 * ]
 * Output: [
 *   ["Ethan","Ethan0@m.co","Ethan4@m.co","Ethan5@m.co"],
 *   ["Gabe","Gabe0@m.co","Gabe1@m.co","Gabe3@m.co"],
 *   ["Hanzo","Hanzo0@m.co","Hanzo1@m.co","Hanzo3@m.co"],
 *   ["Kevin","Kevin0@m.co","Kevin3@m.co","Kevin5@m.co"],
 *   ["Fern","Fern0@m.co","Fern1@m.co","Fern5@m.co"]
 * ]
 *
 * Constraints:
 *
 * * `1 <= accounts.length <= 1000`
 * * `2 <= accounts[i].length <= 10`
 * * `1 <= accounts[i][j].length <= 30`
 * * `accounts[i][0]` consists of English letters.
 * * `accounts[i][j] (for j > 0)` is a valid email.
 */

import { assertArrayIncludes, assertEquals } from "std/assert/mod.ts";

Deno.test("example 1", () => {
  const accounts = [
    ["John", "johnsmith@mail.com", "john_newyork@mail.com"],
    ["John", "johnsmith@mail.com", "john00@mail.com"],
    ["Mary", "mary@mail.com"],
    ["John", "johnnybravo@mail.com"],
  ];
  assertArrayIncludes(
    accountsMerge(accounts),
    [
      ["John", "john00@mail.com", "john_newyork@mail.com", "johnsmith@mail.com"],
      ["Mary", "mary@mail.com"],
      ["John", "johnnybravo@mail.com"],
    ],
  );
});

Deno.test("example 2", () => {
  const accounts = [
    ["Gabe", "Gabe0@m.co", "Gabe3@m.co", "Gabe1@m.co"],
    ["Kevin", "Kevin3@m.co", "Kevin5@m.co", "Kevin0@m.co"],
    ["Ethan", "Ethan5@m.co", "Ethan4@m.co", "Ethan0@m.co"],
    ["Hanzo", "Hanzo3@m.co", "Hanzo1@m.co", "Hanzo0@m.co"],
    ["Fern", "Fern5@m.co", "Fern1@m.co", "Fern0@m.co"],
  ];
  assertArrayIncludes(
    accountsMerge(accounts),
    [
      ["Ethan", "Ethan0@m.co", "Ethan4@m.co", "Ethan5@m.co"],
      ["Gabe", "Gabe0@m.co", "Gabe1@m.co", "Gabe3@m.co"],
      ["Hanzo", "Hanzo0@m.co", "Hanzo1@m.co", "Hanzo3@m.co"],
      ["Kevin", "Kevin0@m.co", "Kevin3@m.co", "Kevin5@m.co"],
      ["Fern", "Fern0@m.co", "Fern1@m.co", "Fern5@m.co"],
    ],
  );
});

Deno.test("failed case", () => {
  const accounts = [
    ["David", "David0@m.co", "David1@m.co"],
    ["David", "David3@m.co", "David4@m.co"],
    ["David", "David4@m.co", "David5@m.co"],
    ["David", "David2@m.co", "David3@m.co"],
    ["David", "David1@m.co", "David2@m.co"],
  ];
  assertArrayIncludes(
    accountsMerge(accounts),
    [
      ["David", "David0@m.co", "David1@m.co", "David2@m.co", "David3@m.co", "David4@m.co", "David5@m.co"],
    ],
  );
});
// @leetup=custom
// @leetup=code

function accountsMerge(accounts) {
  const people = groupByName(accounts);

  const mergedAccounts = [];

  people.forEach((accs, name) => {
    const graph = buildGraph(accs);
    console.log(graph);
    findConnectedComponents(graph).forEach(component => {
      mergedAccounts.push([name, ...component.sort()]);
    })
  });
  return mergedAccounts;
}

function buildGraph(accs) {
  const graph = {};
  accs.forEach((accountEmails) => {
    for (const email of accountEmails) {
      graph[email] ??= new Set();
      for (const adj of accountEmails) {
        graph[email].add(adj);
      }
    }
  });
  return graph;
}

function findConnectedComponents(graph) {
  const visited = new Array(Object.keys(graph).length);
  const components = [];
  for (const email in graph) {
    const currentComponent = [];
    dfs(graph, email, (node) => {
      if (visited[node]) {
        return false;
      }
      currentComponent.push(node);
      visited[node] = true;
      return true;
    });
    if (currentComponent.length > 0) {
      components.push(currentComponent);
    }
  }
  return components;
}

function dfs(graph, startingNode, callback) {
  const stack = [startingNode];
  while (stack.length > 0) {
    const node = stack.pop();
    if (!callback(node)) {
      continue;
    }
    for (const adj of graph[node]) {
      stack.push(adj);
    }
  }
}

function groupByName(accounts) {
  const byName = new Map();
  for (let i = 0; i < accounts.length; i++) {
    let nameGroup = byName.get(accounts[i][0]);
    if (nameGroup === undefined) {
      nameGroup = [];
      byName.set(accounts[i][0], nameGroup);
    }
    const emails = new Set();
    for (let j = 1; j < accounts[i].length; j++) {
      emails.add(accounts[i][j]);
    }
    nameGroup.push(emails);
  }
  return byName;
}
// @leetup=code

File: python3/accounts-merge.py
from typing import List
import itertools


class UnionFind:
    def __init__(self):
        self._parent = {}
        self._size = {}

    def add(self, x):
        if x in self._parent:
            return
        self._parent[x] = x
        self._size[x] = 1

    def find(self, x):
        if x == self._parent[x]:
            return x
        return self.find(self._parent[x])

    def union(self, x, y):
        if x != self._parent[x]:
            return self.union(self._parent[x], y)
        if y != self._parent[y]:
            return self.union(x, self._parent[y])
        if self._size[x] < self._size[y]:
            return self.union(y, x)
        self._parent[y] = x
        self._size[x] += self._size[y]

    def disjoint_sets(self):
        parent_to_children = {}
        for key, value in self._parent.items():
            parent = self.find(key)
            if parent not in parent_to_children:
                parent_to_children[parent] = []
            parent_to_children[parent].append(key)
        return parent_to_children


class Solution:
    """
    >>> Solution().accountsMerge([["John","johnsmith@mail.com","john_newyork@mail.com"],["John","johnsmith@mail.com","john00@mail.com"],["Mary","mary@mail.com"],["John","johnnybravo@mail.com"]])
    [['John', 'john00@mail.com', 'john_newyork@mail.com', 'johnsmith@mail.com'], ['Mary', 'mary@mail.com'], ['John', 'johnnybravo@mail.com']]
    >>> sorted(Solution().accountsMerge([["Gabe","Gabe0@m.co","Gabe3@m.co","Gabe1@m.co"],["Kevin","Kevin3@m.co","Kevin5@m.co","Kevin0@m.co"],["Ethan","Ethan5@m.co","Ethan4@m.co","Ethan0@m.co"],["Hanzo","Hanzo3@m.co","Hanzo1@m.co","Hanzo0@m.co"],["Fern","Fern5@m.co","Fern1@m.co","Fern0@m.co"]]))
    [['Ethan', 'Ethan0@m.co', 'Ethan4@m.co', 'Ethan5@m.co'], ['Fern', 'Fern0@m.co', 'Fern1@m.co', 'Fern5@m.co'], ['Gabe', 'Gabe0@m.co', 'Gabe1@m.co', 'Gabe3@m.co'], ['Hanzo', 'Hanzo0@m.co', 'Hanzo1@m.co', 'Hanzo3@m.co'], ['Kevin', 'Kevin0@m.co', 'Kevin3@m.co', 'Kevin5@m.co']]
    >>> sorted(Solution().accountsMerge([["Ethan","Ethan1@m.co","Ethan2@m.co","Ethan0@m.co"],["David","David1@m.co","David2@m.co","David0@m.co"],["Lily","Lily0@m.co","Lily0@m.co","Lily4@m.co"],["Gabe","Gabe1@m.co","Gabe4@m.co","Gabe0@m.co"],["Ethan","Ethan2@m.co","Ethan1@m.co","Ethan0@m.co"]]))
    [['David', 'David0@m.co', 'David1@m.co', 'David2@m.co'], ['Ethan', 'Ethan0@m.co', 'Ethan1@m.co', 'Ethan2@m.co'], ['Gabe', 'Gabe0@m.co', 'Gabe1@m.co', 'Gabe4@m.co'], ['Lily', 'Lily0@m.co', 'Lily4@m.co']]
    """

    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        name_by_email = {}
        union_find = UnionFind()
        for accs in accounts:
            name = accs[0]
            first_email = accs[1]
            union_find.add(first_email)
            name_by_email[first_email] = name
            for i in range(2, len(accs)):
                name_by_email[accs[i]] = name
                union_find.add(accs[i])
                union_find.union(first_email, accs[i])
        merged = []
        for first_email, emails in union_find.disjoint_sets().items():
            name = name_by_email[first_email]
            merged.append([name, *sorted(emails)])
        return merged
